### PR TITLE
Refactor CLI flag parsing

### LINKED ISSUE #1987: Refactor CLI flag parsing
I have been putting this off long enough - and heard enough feedback from people that I’m ready to relent on some particular ideas I had around CLI parsing.

In particular, let’s move to the model the flags after the script are passed to Deno.args. Let’s allow for parameters to arguments so we can support evaluating oneliners (-e) and import maps. And we should consider making “deno —fmt” into “deno fmt”
Ideally there’s a rust crate which we can offload this onto? Anyone have proposals for such a crate? Anyone want to take this task on?

### PR DESCRIPTION
**Pull request by bartlomieju**
This PR closes #1987

Blocked by #2017 

I've started out by swapping out `getopts` for `clap` parsing. Right now there's some strange code to provide backwards compatibility.

@ry I see that V8 flag logic is pretty obscure, I verified that we could leverage such syntax:
```bash
deno --v8-flags=--predictable,--log-instruction-status=1,--single-threaded=/asdf/sadf --allow-all some_script.ts
```
Unfortunately it has to be written as `--v8-flags=`. Let me know what you think.

### CODE REVIEW

**Comment by ry** [comment]
@bartlomieju Cool - the V8 flag situation isn't ideal, but I'm ok with moving forward with it anyway.

@piscisaureus is working on rebasing #2017 - we've been doing a lot of reorg in the tree so it might take a day or two.

**Comment by bartlomieju** [comment]
Thanks fine for me. I started work on subcommands (`deno fmt`, etc.), it looks promising. 

I expect some turmoil re: Deno.args - quite a lot of files mix Deno flags with custom args, but we can sort it out step-by-step over a few PRs.

**Comment by ry** [comment]
@bartlomieju #2017 was landed in master - it didn't take as long as originally thought.

File: cli/flags.rs [hunk 1]
```diff
@@ -1,7 +1,7 @@
 // Copyright 2018-2019 the Deno authors. All rights reserved. MIT license.
+use clap::{App, AppSettings, Arg, ArgMatches};
+use crate::ansi;
 use deno::v8_set_flags;
-use getopts;
-use getopts::Options;
```
Comment (ry) [inline cli/flags.rs]: getopts should be removed from Cargo.toml and cli/BUILD.gn too

**Comment by ry** [comment]
Screen Shot 2019-04-02 at 12 32 33 PM

Is there a way to tune down the color? Especially the green seems fairly unreadable.

File: cli/flags.rs (hunk 1 above)
Comment (bartlomieju) [reply cli/flags.rs]: Will do

**Comment by bartlomieju** [comment]
@ry I agree it's a bit too bright, I'll have to check if we can change that.

Right now the changes are more-or-less compatible with our current way of handling args. WRT comment in #1987 I understand we want to have:
```bash
# new way or parsing args, deno only parses flags appearing 
# before <entry_point> argument, all the rest is forwarded to Deno.args
deno --allow-env --allow-net entry_point.ts arg1 arg2 --another-flag
```

```ts
// entry_point.ts
console.log(Deno.args);
```
This will result in:
`Deno.args: [ "entry_point.ts", "arg1", "arg2", "--another-flag"]`


Formatting will be done via subcommand:
```bash
# before
deno --fmt <file>... 
# after 
deno fmt <file>...
```

Last but not least V8 flags handling:
```bash
# help for V8 flags
# before
deno --v8-options
# after 
deno --v8-help

# pass flags to V8
deno --v8-options=--predictable,--log-instruction-status=1,--single-threaded --allow-all -D
```

Does that good for you @ry?

**Comment by ry** [comment]
`--v8-options` is very familiar to those who use Node, so I'd like to keep it printing help if possible. What if `--v8-options` prints help when no arguments are given?

**Comment by bartlomieju** [comment]
@ry I tried to do that, but clap requires to use `Arg.takes_value(true)` to be able to get arg's value. So it looks like we can't use it as a flag and take option at the same time.

**Comment by ry** [comment]
@bartlomieju Ok. We can revisit later.

I will wait for your update to BUILD.gn and Cargo.toml to remove getopts dep. Otherwise I think this is ready?

**Comment by bartlomieju** [comment]
@ry Actually not, I need to implement all the things from previous comment, unless to want to land it now and do further changes in next PR.

**Comment by bartlomieju** [comment]
@ry I believe CI is false positively green. In the log there is no sign of `test_set_flags_*` tests, and it seems they do not pass locally. Debugging..

**Comment by bartlomieju** [comment]
@ry after going back and forth with parsing args I have to propose to go with this method:
```bash
USAGE: deno [FLAGS] <script> -- <script_args>...
```
Which requires passing script arguments after `--`. This approach proved to be the most robust one. 

Doing this:
```bash
USAGE: deno [FLAGS] <script> <script_args>...
```
turns out to be huge PITA, and causing lot of quirks... Eg: `deno --allow-net gist.ts --title X` throws an error.

Also I wasn't yet able to force `clap` to require flags before `<script>` which means that you can still do both: `deno --allow-net script.ts` and `deno script.ts --allow-net`

**Comment by bartlomieju** [comment]
Regarding changing help colors: clap-rs/clap/#836

**Comment by hayd** [comment]
I wonder if there's a clever way to use [AllowExternalSubcommands](https://docs.rs/clap/2.24.2/clap/enum.AppSettings.html#variant.AllowExternalSubcommands) (to fall to a "script" subcommand).

File: cli/flags.rs [hunk 2]
```diff
@@ -51,129 +40,177 @@ Environment variables:
...
-        if matches.opt_present("fmt") {
-          flags.fmt = true;
-        }
-
-        if !matches.free.is_empty() {
-          rest.extend(matches.free);
-        }
-      }
-    }
+fn set_recognized_flags(matches: ArgMatches, flags: &mut DenoFlags) {
```
Comment (hayd) [inline cli/flags.rs]: This could be `impl From<ArgMatches> for DenoFlags`

**Comment by ry** [comment]
> turns out to be huge PITA, and causing lot of quirks... Eg: deno --allow-net gist.ts --title X throws an error

Dang that sucks. I thought this was one of clap’s features (but now that read back on their documents, I see I had confused it.)

Also the 2 year old open issue for color support isn’t very heartening. 

@bartlomieju I haven’t had a chance to really dig into your code yet. What’s your feeling so far? Is this going to be better? Do you think clap is the right choice?

**Comment by hayd** [comment]
I think treating foo.ts as an  ExternalSubcommands ought to work without --.

See re clap maintainer: https://www.reddit.com/r/rust/comments/b575hi/comment/ejdawiv

File: cli/flags.rs (hunk 2 above)
Comment (bartlomieju) [reply cli/flags.rs]: Sure, haven't done refactor yet, rather switching the engine leaving the rest as is. Thanks :+1:

**Comment by bartlomieju** [comment]
> I wonder if there's a clever way to use [AllowExternalSubcommands](https://docs.rs/clap/2.24.2/clap/enum.AppSettings.html#variant.AllowExternalSubcommands) (to fall to a "script" subcommand). Edit: to avoid the need for --.

@hayd this is promising, I'll try that.


@ry I have to say `clap` is really handy, and it makes it easy to refactor `main.rs` further to make it true CLI entry point, but... those quirks with `--` and coloring issue (it seems coloring doesn't work on Windows at all) make me a little sad. I have no problem to try other solution (maybe [docopt.rs](https://github.com/docopt/docopt.rs)?) to compare results.

**Comment by bartlomieju** [comment]
> I think treating foo.ts as an ExternalSubcommands ought to work without --.
> 
> See re clap maintainer: https://www.reddit.com/r/rust/comments/b575hi/comment/ejdawiv

@hayd that's a bullseye!

Now we can have:
```bash
$ deno [FLAGS] [script] [script_args]
```
It will require `[FLAGS]` to be passed before script, so `deno --allow-all script.ts` will work, but `deno script.ts --allow-all` will not. 

@ry that leaves us only with `help` coloring problem, can we live with that for now waiting for `clap v3`?

**Comment by ry** [comment]
@bartlomieju good to hear! Can you just turn off color entirely?

**Comment by bartlomieju** [comment]
Current `deno --help` output:

Zrzut ekranu 2019-04-6 o 17 10 39

I can mimic clap's coloring for environment variables section.

@ry PTAL, most of the changes are already there, there are two more things I want to do:
- implement @hayd's suggestion - `impl From<ArgMatches> for DenoFlags`
- refactor `main.rs` and `set_flags` - move `clap_app` to `main.rs`

**Comment by bartlomieju** [comment]
@ry you were faster 😅 no problem, turning color off

File: cli/main.rs [hunk 3]
```diff
@@ -82,24 +84,20 @@ fn main() {
...
-  }
-
   log::set_max_level(if flags.log_debug {
     LevelFilter::Debug
   } else {
     LevelFilter::Warn
   });
 
   if flags.fmt {
+    println!("flags.fmt! {:?}", rest_argv);
```
Comment (ry) [inline cli/main.rs]: Remove debug output

File: core/flags.rs [hunk 4]
```diff
@@ -3,19 +3,15 @@
...
-use std::ffi::CStr;
 use std::ffi::CString;
-use std::mem;
 use std::vec::Vec;
 
 /// Pass the command line arguments to v8.
 /// Returns a vector of command line arguments that V8 did not understand.
 /// Translates --v8-options into a --help flag for V8.
-pub fn v8_set_flags(args: Vec<String>) -> Vec<String> {
+pub fn v8_set_flags(args: Vec<String>) {
```
Comment (ry) [inline core/flags.rs]: I don't think you need to change this interface. You can just ignore the return value in CLI ?

Comment (bartlomieju) [reply core/flags.rs]: You're right, reverted

**Comment by ry** [comment]
Please update manual.md with the new help text

**Comment by ry** [comment]
Can you change "--info" into "deno info" too?

**Comment by ry** [comment]
Is there a test for Deno.args ? I think not. Since the arguments are changing around, I think you should add something that checks the args. It could be as simple as adding a file to tests/ which has a bunch of args and prints them out.

**Comment by bartlomieju** [comment]
Thanks @ry! I'll clean up a few more things in follow up PR

File: cli/flags.rs [hunk 5]
```diff
@@ -51,129 +39,217 @@ Environment variables:
...
+      // TODO(bartlomieju): it still relies on `is_present("fmt")` check
+      // in `set_recognized_flags`
+      let files: Vec<String> = fmt_match
+        .values_of("files")
+        .unwrap()
+        .map(String::from)
+        .collect();
+      rest.extend(files);
+    }
+    (script, Some(script_match)) => {
```
Comment (hayd) [inline cli/flags.rs]: Could this somehow match only if endswith .ts or .js.

At the moment, typos aren't caught correctly e.g. --allow-ne

Comment (bartlomieju) [reply cli/flags.rs]: it behaves exactly like previous version, we can change it further. I'll put up another PR tonight with more refactors.

Comment (hayd) [reply cli/flags.rs]: "it behaves exactly like previous version" This isn't quite true (though I think it's a big improvement!) previously --allow-ne would be silently passed through to the script as an arg, now it's treated as a potential script (correctly thowing an error, but a confusing one).

Comment (bartlomieju) [reply cli/flags.rs]: @hayd can you paste an example?

Comment (hayd) [reply cli/flags.rs]: `deno --allow-ne foo.ts`

Comment (hayd) [reply cli/flags.rs]: Huh! Now I try this again, this isn't the case! My mistake! 👍

Comment (hayd) [reply cli/flags.rs]: This (new behavior) is _really_ nice:
```
error: Found argument '--allow-ne' which wasn't expected, or isn't valid in this context
	Did you mean --allow-net?

USAGE:
    deno --allow-net

For more information try --help
```

**Pull request merged**